apps/aecore/lib/aecore/contract/contract_state_tree.ex

```diff
@@ -0,0 +1,118 @@
+defmodule Aecore.Contract.ContractStateTree do
+  @moduledoc """
+  Top level contract state tree.
+  """
+  alias Aecore.Chain.Identifier
+  alias Aecore.Contract.Contract
+  alias Aeutil.PatriciaMerkleTree
+  alias Aeutil.Serialization
+
+  @contract_key_size 32
+
+  @type hash :: binary()
+  @type contracts_state() :: Trie.t()
+
+  @spec init_empty() :: contracts_state()
+  def init_empty do
+    PatriciaMerkleTree.new(:contracts)
+  end
+
+  @spec insert_contract(contracts_state(), map()) :: contracts_state()
+  def insert_contract(contract_tree, contract) do
```

Comment:
There are a few places like this where you could pattern match a value out of the `%Contract` struct. In this case the `contract.id`